conn.go

```diff
@@ -72,3 +74,96 @@ func (c *Conn) newError(apiName string, handle interface{}) error {
 	}
 	return err
 }
+
+// QueryContext implements the driver.QueryerContext interface.
+// As per the specifications, it honours the context timeout and returns when the context is cancelled.
+// When the context is cancelled, it first cancels the statement, closes it, and then returns an error.
+func (c *Conn) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	// Prepare a query
+	os, err := c.PrepareODBCStmt(query)
+	if err != nil {
+		return nil, err
+	}
+
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Execute the statement
+	rowsChan := make(chan driver.Rows)
+	defer close(rowsChan)
+	errorChan := make(chan error)
+	defer close(errorChan)
+
+	if ctx.Err() != nil {
+		os.closeByStmt()
+		return nil, ctx.Err()
+	}
+
+	go c.wrapQuery(ctx, os, dargs, rowsChan, errorChan)
+
+	var finalErr error
+	var finalRes driver.Rows
+
+	select {
+	case <-ctx.Done():
+		// Context has been cancelled or has expired, cancel the statement
+		if err := os.Cancel(); err != nil {
+			finalErr = err
+			break
+		}
+
+		// The statement has been cancelled, the query execution should eventually fail now.
+		// We wait for it in order to avoid having a dangling goroutine running in the background
+		<-errorChan
+		finalErr = ctx.Err()
+	case err := <-errorChan:
+		finalErr = err
+	case rows := <-rowsChan:
+		finalRes = rows
+	}
+
+	// Close the statement
+	os.closeByStmt()
+	os = nil
+
+	return finalRes, finalErr
+}
+
+// wrapQuery is following the same logic as `stmt.Query()` except that we don't use a lock
+// because the ODBC statement doesn't get exposed externally.
+func (c *Conn) wrapQuery(ctx context.Context, os *ODBCStmt, dargs []driver.Value, rowsChan chan<- driver.Rows, errorChan chan<- error) {
+	if err := os.Exec(dargs, c); err != nil {
+		errorChan <- err
+		return
+	}
+
+	if err := os.BindColumns(); err != nil {
+		errorChan <- err
+		return
+	}
+
+	os.usedByRows = true
```

Comment:
Regarding [this](https://github.com/alexbrainman/odbc/pull/175#discussion_r845662123) comment, there is no race condition here.
If line 128 is reached, one of the following has happened:
1. The context got canceled -> os.Cancel() gets called, and then the main thread waits for wrapQuery to return an error on errorChan. In this case, line 128 is not reached until after wrapQuery returns.
2. wrapQuery returns an error on the errorChan. Same as above.
3. wrapQuery returns some rows.

In all three cases, wrapQuery returns before line 128 is reached. Hence there is no race condition, and no need for a lock.